python/tank/context.py

```diff
@@ -624,37 +623,105 @@ def _fields_from_template_tree(self, template, fields, entities):
+                    #
+                    # We are searching for fields for 'Shot_B' that has a broken entry in the path cache so the fields 
+                    # returned for each level of the template will be:
+                    #    /{Sequence}                 -> {"Sequence":"Seq_001"} <- Correct
+                    #    /{Sequence}/{Shot}          -> {}                     <- entry not found for Shot_B matching 
+                    #                                                             the template
+                    #    /{Sequence}/{Shot}/{Step}   -> {"Sequence":"Seq_001", <- Correct
+                    #                                    "Shot":"Shot_A",      <- Wrong!
+                    #                                    "Step":"Lighting"}    <- Correct
+                    #
+                    # The final fields returned are:
+                    #
+                    #     {"Sequence":"Seq_001",
+                    #      "Shot":"Shot_A",
+                    #      "Step":"Lighting"}
+                    #
+                    # The wrong Shot (Shot_A) is returned and not caught because the code only tested that the Step
+                    # entity matches and just assumes that the rest is correct - this isn't the case when there is
+                    # a one-to-many relationship between entities!
+                    #
+                    # Therefore, we need to validate that we didn't find any entity fields that we should have found
+                    # previously/higher up in the template definition.  If we did then the entries that were found 
+                    # may not be correct so we have to discard them!
+                    found_mismatching_field = False
+                    for field_name, field_value in entity_fields.iteritems():
+                        if field_name in all_fields:
+                            # already validated/processed this field so can skip
+                            continue
+                        elif field_name == key_name:
+                            # the found field matches the entity we were searching for so must be valid!
+                            found_fields[field_name] = field_value
+                        elif field_name in entities:
+                            # We found an entity type that we should have found before (in a previous/shorter 
+                            # template).  This means we can't trust any other fields that were found as they
+                            # may belong to a completely different entity/path! 
+                            found_mismatching_field = True
+
+                    if not found_mismatching_field:
+                        # all fields are ok so we can add them all to the full list of fields :)
+                        found_fields.update(entity_fields)
+
+                    # update all fields with any found fields
+                    all_fields.update(found_fields)
```

Comment:
wondering if all_fields and found_fields could somehow be the same thing? My head hurts at this point :)